<a href="https://colab.research.google.com/github/623020514-5/data_viz_2022/blob/main/Final_DATAVIZ.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

จากโจทย์
ให้นักศึกษาสร้าง Dashboard แสดงรายงานเกี่ยวกับข้อมูลการตรวจโรค(ov) จาก ไฟล์ alldb.csv และ hospital_list.xlsx
( ดาวน์โหลดจาก https://drive.google.com/drive/folders/1D_Z0fjTki6U6ECbgh2V3wBOSdOXS81YT?fbclid=IwAR2TOFCt16PG2k9xeKIA7jnTEm63uTzd3nRacpzhJEp3d3-6QCtWz3tQskw )


วิธีทำ
#**เรียกใช้ฟังก์ชั่น**


In [28]:
import pandas as pd #เพื่อใช้ทำงานกับข้อมูลที่เป็นรูปแบบตาราง
import os           #เพื่อจัดการกับไฟล์ และ ตัว path ที่เราจะนำเข้า
import matplotlib
from matplotlib import pyplot as plt

#**นำเข้าไฟล์**

In [29]:
from google.colab import drive#เชื่อมข้อมูล เพราะเราจะทำการนำเข้าข้อมูล โดยให้ดึงมาจาก google drive ของเรา
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [30]:
path = '/content/drive/My Drive/FinalDataViz'
#โดยเราจะทำการระบุให้ชี้ไปที่ โฟนเดอร์ที่เราเซฟไฟล์ไว้ 

##**ตารางที่ 1**
>date เป็น YYYY-MM-DD

In [31]:
data1 = os.path.join(path,'alldb.csv') 
data1 = pd.read_csv(data1,parse_dates=['date']) #date เป็น YYYY-MM-DD
data1

Unnamed: 0           user                                    RowKey  \
0               0           test                                       NaN   
1               1          04857      136FD222-FD9F-4E54-AE1F-7282A276B085   
2               2            NaN  1E6D3D62-4C8A-4F8E-90D1-0DC5BC2752DF.png   
3               3  3450101576642      2D331FB0-CBA2-4586-A5B0-CB3CF44DD9FE   
4               4  3450101576642      45197AF7-7CB4-470C-9ABB-37C092BD6B21   
...           ...            ...                                       ...   
40030       40030            NaN      6B5FA20F-190E-4021-8547-C12308024C6C   
40031       40031            NaN      F48C51C0-A345-40D6-93F4-46213267E1B1   
40032       40032            NaN      A986BDAA-6D7C-4017-8E62-A3C104E946B9   
40033       40033            NaN                            bbuogxulur.png   
40034       40034            NaN                            zttpqogiso.png   

      care       date   ov             phone   site      time  \
0      NaN        NaT  NaN               NaN    NaN       NaN   
1      NaN 2022-02-24  NaN        iPhone10,4  05310  10:50:32   
2      NaN        NaT  NaN               NaN    NaN       NaN   
3      NaN 2022-02-23    0        iPhone13,4  05087  09:59:47   
4      NaN 2022-02-23  NaN        iPhone13,4  05087  09:44:06   
...    ...        ...  ...               ...    ...       ...   
40030  NaN        NaT  NaN        iPhone13,4  01234       NaN   
40031  NaN        NaT    2        iPhone10,5  01234       NaN   
40032  NaN        NaT    1        iPhone13,4  01234       NaN   
40033  NaN        NaT  NaN  Samsung SM-A115F  01234       NaN   
40034  NaN        NaT    1  Samsung SM-A115F  01234       NaN   

                 urine  pid            PartitionKey                  title  \
0                  NaN  NaN                     NaN  data is not completed   
1                  NaN  NaN                     NaN  data is not completed   
2                  NaN  NaN                     NaN  data is not completed   
3      ปกติ (เหลืองใส)  NaN                     NaN  data is not completed   
4      ปกติ (เหลืองใส)  NaN                     NaN  data is not completed   
...                ...  ...                     ...                    ...   
40030              NaN    5     เกษศิรินทร์  ไชยลาภ  data is not completed   
40031  ปกติ (เหลืองใส)    5  เถลิงเกียรติ เสริฐเลิศ  data is not completed   
40032  ปกติ (เหลืองใส)    1                    เสรี  data is not completed   
40033              NaN    1                    เสรี  data is not completed   
40034  ปกติ (เหลืองใส)    1                    เสรี  data is not completed   

                         age  
0      data is not completed  
1      data is not completed  
2      data is not completed  
3      data is not completed  
4      data is not completed  
...                      ...  
40030  data is not completed  
40031  data is not completed  
40032  data is not completed  
40033  data is not completed  
40034  data is not completed  

[40035 rows x 14 columns]

##**ตวจสอบขนาด**

In [32]:
data1.shape

(40035, 14)

เนื่องจากข้อมูลมีจำนวนมาก แต่เราไม่ได้ใช้ทั้งหมด เพื่อง่ายต่อการทำ prepocessing ดังนั้นจะตัดมาแค่ข้อมูลที่ต้องการใช้จริงๆ ตามโจทย์


In [33]:
data2 = data1[['user','RowKey','date','phone','site','time','urine','title','ov','PartitionKey','age']] 
data2

user                                    RowKey       date  \
0               test                                       NaN        NaT   
1              04857      136FD222-FD9F-4E54-AE1F-7282A276B085 2022-02-24   
2                NaN  1E6D3D62-4C8A-4F8E-90D1-0DC5BC2752DF.png        NaT   
3      3450101576642      2D331FB0-CBA2-4586-A5B0-CB3CF44DD9FE 2022-02-23   
4      3450101576642      45197AF7-7CB4-470C-9ABB-37C092BD6B21 2022-02-23   
...              ...                                       ...        ...   
40030            NaN      6B5FA20F-190E-4021-8547-C12308024C6C        NaT   
40031            NaN      F48C51C0-A345-40D6-93F4-46213267E1B1        NaT   
40032            NaN      A986BDAA-6D7C-4017-8E62-A3C104E946B9        NaT   
40033            NaN                            bbuogxulur.png        NaT   
40034            NaN                            zttpqogiso.png        NaT   

                  phone   site      time            urine  \
0                   NaN    NaN       NaN              NaN   
1            iPhone10,4  05310  10:50:32              NaN   
2                   NaN    NaN       NaN              NaN   
3            iPhone13,4  05087  09:59:47  ปกติ (เหลืองใส)   
4            iPhone13,4  05087  09:44:06  ปกติ (เหลืองใส)   
...                 ...    ...       ...              ...   
40030        iPhone13,4  01234       NaN              NaN   
40031        iPhone10,5  01234       NaN  ปกติ (เหลืองใส)   
40032        iPhone13,4  01234       NaN  ปกติ (เหลืองใส)   
40033  Samsung SM-A115F  01234       NaN              NaN   
40034  Samsung SM-A115F  01234       NaN  ปกติ (เหลืองใส)   

                       title   ov            PartitionKey  \
0      data is not completed  NaN                     NaN   
1      data is not completed  NaN                     NaN   
2      data is not completed  NaN                     NaN   
3      data is not completed    0                     NaN   
4      data is not completed  NaN                     NaN   
...                      ...  ...                     ...   
40030  data is not completed  NaN     เกษศิรินทร์  ไชยลาภ   
40031  data is not completed    2  เถลิงเกียรติ เสริฐเลิศ   
40032  data is not completed    1                    เสรี   
40033  data is not completed  NaN                    เสรี   
40034  data is not completed    1                    เสรี   

                         age  
0      data is not completed  
1      data is not completed  
2      data is not completed  
3      data is not completed  
4      data is not completed  
...                      ...  
40030  data is not completed  
40031  data is not completed  
40032  data is not completed  
40033  data is not completed  
40034  data is not completed  

[40035 rows x 11 columns]

##**ตวจสอบว่ามี NaN หรือ Missing หรือไม่**

In [34]:
data2.isnull().any() 
#.isnull() เพื่อระบุว่านิพจน์มีค่า Null หรือไม่
#.any() เพื่อดูรายละเอียดการเปรียบเทียบข้อมูลให้ออกมาในรูป true false

user            True
RowKey          True
date            True
phone           True
site            True
time            True
urine           True
title           True
ov              True
PartitionKey    True
age             True
dtype: bool

จากผลเราจะพิจารณาจากการอ่านค่า True และ False ถ้า True คือ มีค่า missing ดังนั้น จะเห็นได้ว่าข้อมูลมีค่า missing เกือบในทุกคอลัมน์ ดังนั้น สามารถอนุมานได้เลยว่า ข้อมูลมีค่า missing จำนวนมาก จึงไม่เหมาะกับตัดข้อมูล หรือ .dropna
จึงพิจารณาให้ใช้วิธีแทนค่าเฉพาะคอลัมน์ข้อมูลที่เราต้องการใช้ 

##**ลองใช้ .dropna ซึ่งจะใช้ตัดแถวข้อมูลที่มีค่า missing**

In [35]:
data_drop = data2.dropna() #ใช้ dropna เพื่อต้องการลบข้อมูลทั้งแถวออกเมื่อมีคอลัมน์ใดคอลัมน์หนึ่งเป็น NaN
data_drop

user                                RowKey       date  \
186    45-161  D2CABB8C-533B-450A-AFA6-B693749CC1C6 2022-02-23   
187    45-161  57A5F06C-C840-49EA-9514-A30D3C897CC8 2022-02-23   
188     27529  ADC54E82-4225-4DE6-89A6-A7E556C15FB1 2022-02-25   
191     12015  00859DB9-1BD9-4825-B246-3DC25EFD1924 2022-03-30   
192     91811  1376ACA2-AFA5-4457-967A-D2FCF908E6A5 2022-03-30   
...       ...                                   ...        ...   
40013   ttest                        crqcufokmn.png 2022-02-21   
40014   ttest                        gzxgflnowr.png 2022-02-21   
40016     xxx                        jytlopevha.png 2022-02-17   
40022     yyy                        ivccsksowb.png 2022-02-17   
40023     zzz                        psudrnccsh.png 2022-02-17   

                  phone   site      time            urine  \
186          iPhone11,8  04312  09:45:28  ปกติ (เหลืองใส)   
187          iPhone11,8  04312  10:06:45  ปกติ (เหลืองใส)   
188          iPhone13,1  04999  14:22:22  ปกติ (เหลืองใส)   
191          iPhone12,3  06307  14:46:08  ปกติ (เหลืองใส)   
192          iPhone14,3  05328  15:27:45  ปกติ (เหลืองใส)   
...                 ...    ...       ...              ...   
40013  Samsung SM-G998B   test  14:23:45             ขุ่น   
40014  Samsung SM-G998B   test  14:23:43             ขุ่น   
40016  Samsung SM-N950F    xxx  00:01:06  ปกติ (เหลืองใส)   
40022  Samsung SM-N950F    yyy  08:41:05   มีเม็ดเลือดแดง   
40023  Samsung SM-N950F    zzz  08:41:04          มีตะกอน   

                       title ov PartitionKey                    age  
186    data is not completed  0       0-1528  data is not completed  
187    data is not completed  0       0-1544  data is not completed  
188    data is not completed  0         0000  data is not completed  
191                   นางสาว  4        00001                     47  
192                      นาง  4        00001                     41  
...                      ... ..          ...                    ...  
40013  data is not completed  2         x001  data is not completed  
40014  data is not completed  2         x001  data is not completed  
40016  data is not completed  3          xxx  data is not completed  
40022  data is not completed  2          yyy  data is not completed  
40023  data is not completed  2          zzz  data is not completed  

[38135 rows x 11 columns]

In [36]:
removes = data2.shape[0] - data_drop.shape[0]
print(f'size before drop = {data2.shape[0]}')
print(f'size after drop = {data_drop.shape[0]}')
print(f'% from drop = {(removes/data2.shape[0])*100}')

size before drop = 40035
size after drop = 38135
% from drop = 4.745847383539403


ดังนั้น จากการ dropna จะเห็นได้ว่าค่าข้อมูลหายไปเพียงแค่ 4.75 เปอร์เซ็นเท่านั้น ซึ่งอยู่ในช่วงที่ยอมรับได้ จึงสามารถใช้ค่าข้อมูลจากการ dropna ได้ 

>เปลี่ยนชนิดข้อมูล

In [37]:
import numpy as np

เปลี่ยนค่าที่เราต้องการเป็น int หรือเลขจำนวนเต็มก่อน โดยกำหนด error='coerce' จะแปลงค่าที่ไม่ใช่ตัวเลขเป็น NaN

In [38]:
data_drop.dtypes

user                    object
RowKey                  object
date            datetime64[ns]
phone                   object
site                    object
time                    object
urine                   object
title                   object
ov                      object
PartitionKey            object
age                     object
dtype: object

In [39]:
data_drop['user'] = pd.to_numeric(data_drop['user'], errors='coerce')
data_drop['site'] = pd.to_numeric(data_drop['site'], errors='coerce')
data_drop['ov'] = pd.to_numeric(data_drop['ov'], errors='coerce')
data_drop['PartitionKey'] = pd.to_numeric(data_drop['PartitionKey'], errors='coerce')
data_drop['age'] = pd.to_numeric(data_drop['age'], errors='coerce')

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documen

In [40]:
data_drop.dtypes

user                   float64
RowKey                  object
date            datetime64[ns]
phone                   object
site                   float64
time                    object
urine                   object
title                   object
ov                     float64
PartitionKey           float64
age                    float64
dtype: object